opts/mount.go

```diff
@@ -112,6 +113,33 @@ func (m *MountOpt) Set(value string) error {
 			if err != nil {
 				return fmt.Errorf("invalid value for %s: %s", key, val)
 			}
+			logrus.Warn("bind-nonrecursive is deprecated, use bind-recursive=disabled instead")
+		case "bind-recursive":
+			valS := val
+			// Allow boolean as an alias to "enabled" or "disabled"
+			if b, err := strconv.ParseBool(valS); err == nil {
+				if b {
+					valS = "enabled"
+				} else {
+					valS = "disabled"
+				}
+			}
+			switch valS {
+			case "enabled": // read-only mounts are recursively read-only if Engine >= v25 && kernel >= v5.12
+				// NOP
+			case "disabled": // alias of bind-nonrecursive=true
+				bindOptions().NonRecursive = true
+			case "readonly-classic": // conforms to Docker v24; read-only mounts are recursively mounted but not recursively read-only
+				bindOptions().ReadOnlyNonRecursive = true
+				mount.ReadOnly = true
+			case "readonly": // force recursively read-only, or raise an error
+				bindOptions().ReadOnlyForceRecursive = true
+				bindOptions().Propagation = mounttypes.PropagationRPrivate
+				mount.ReadOnly = true
```

Comment:
I think we're getting close (sorry had to jump out earlier as I had some calls); if we consider the `bind-recursive` option to be controlling options for _recursive mounts_ (and not for the "main" mount), should we update `mount.Readonly` at all here?

My thinking (please reply if I'm completely off the mark ❤️);

- `readonly` is the "overall" option; when enabled, by default it makes both the "main" mount, and any nested mounts read-only (i.e. "best-effort recursive-readonly"
- `bind-recursive=disabled` disables nested mounts
- `bind-recursive=readonly` forces nested mounts to be read-only
    - 👆 this can also be done separate from `readonly`, so `--mount type=bind,source=host-path,target=container-path,bind-recursive=readonly` would then mean: the main mount is read-write, but nested mounts (if any) would be mounted read-only
    - 👆 ❓ is that combination technically possible?
- which leaves `readonly-classic` if we don't set `mount.ReadOnly` there, would that make it the same as `enabled` ? (i.e.; nested mounts are mounted, but "read-write") ? If that's the case, then _maybe_ we don't need that option 🤔

Reply:
`mount.ReadOnly = true` is set here, as `mount.ReadOnly == false && bindOptions().ReadOnlyXXX = true` doesn't seem a valid configuration.

Reply:
That's the part I was wondering; is it (technically) possible to make the "main" mount (so `target`) read-write, but making _nested_ mounts read-only?

Reply:
Basically; my thinking here is to make nested mounts "inherit" the `readonly` option as a default, with the option to differentiate (through the `bind-recursive` options). In that case it's (I _think_) easier to comprehend (by default, they'll all be attempted the same, but if I need them to be different, I can control that through this option)

Reply:
> That's the part I was wondering; is it (technically) possible to make the "main" mount (so `target`) read-write, but making _nested_ mounts read-only?

Theoretically possible by scanning procfs, but I don't think we want to implement it